tuner/libtuner.py

```diff
@@ -0,0 +1,1302 @@
+
+    return all_device_ids
+
+
+def parse_devices(devices_str: str) -> list[str]:
+    """
+    Parse a comma-separated list of device IDs e.g.:
+    --devices=hip://0,local-sync://default -> ["hip://0", "local-sync://default"]).
+    """
+    devices = [device.strip() for device in devices_str.split(",")]
+    for device in devices:
+        if "://" not in device or not device:
+            handle_error(
+                condition=True,
+                msg=f"Invalid device list: {devices_str}. Error: {ValueError()}",
+                error_type=argparse.ArgumentTypeError,
+            )
+    return devices
+
+
+def validate_devices(user_devices: list[str]) -> None:
+    """Validates the user provided devices against the devices extracted by the IREE Runtime"""
+    user_drivers = extract_driver_names(user_devices)
+
+    available_devices = fetch_available_devices(list(user_drivers))
+
+    for device in user_devices:
+        handle_error(
+            condition=(device not in available_devices),
+            msg=f"Invalid device specified: {device}\nFetched available devices: {available_devices}",
+            error_type=argparse.ArgumentError,
+            exit_program=True,
+        )
+
+
+class ExecutionPhases(str, Enum):
+    dont_stop = ""
+    generate_candidates = "generate-candidates"
+    compile_dispatches = "compile-dispatches"
+    benchmark_dispatches = "benchmark-dispatches"
+    compile_models = "compile-models"
+    benchmark_models = "benchmark-models"
+
+
+def parse_arguments() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(description="Autotune script")
+
+    # Required arguments
+    parser.add_argument(
+        "input_file", type=Path, help="Path to the input benchmark file (.mlir)"
+    )
+
+    # General options
+    parser.add_argument(
+        "--verbose", "-v", action="store_true", help="Enable verbose output to stdout"
+    )
+    parser.add_argument(
+        "--devices",
+        type=parse_devices,
+        default=DEFAULT_DEVICE_LIST,
+        help="Comma-separated list of device IDs (e.g., --devices=hip://,hip://GPU-UUID).",
+    )
+    parser.add_argument(
+        "--max-cpu-workers",
+        type=int,
+        default=DEFAULT_MAX_CPU_WORKERS,
+        help=f"Max number of workers for CPU-bounding tasks (default: {DEFAULT_MAX_CPU_WORKERS}, the number of CPUs in current system)",
+    )
+    parser.add_argument(
+        "--stop-after",
+        choices=[x.value for x in ExecutionPhases],
+        default=ExecutionPhases.dont_stop,
+        help="Stop execution after specified phase",
+    )
+    parser.add_argument(
+        "--num-model-candidates",
+        help="Maximum number of stage 2 candidates",
+        type=int,
+        default=50,
+    )
+    parser.add_argument(
+        "--dry-run",
+        action="store_true",
+        help="Do not attempt to run any modules or initialize the IREE runtime",
+    )
+
+    # candidate_gen.tune() options
+    parser.add_argument(
```

Comment:
You can use argument groups for these: https://docs.python.org/3/library/argparse.html#argument-groups

Reply:
thanks!